File: sdk/js/tests/test_client.js
import test from 'node:test';
import assert from 'node:assert';
import { Lychee, LycheeError } from '../src/index.js';

test('Lychee Client Initialization', (t) => {
  const client = new Lychee();
  assert.strictEqual(client.baseUrl, 'http://localhost:11434');

  const clientCustom = new Lychee('http://192.168.1.50:11434/');
  assert.strictEqual(clientCustom.baseUrl, 'http://192.168.1.50:11434');
});

test('Lychee Client methods success flows', async (t) => {
  const client = new Lychee();
  const originalFetch = globalThis.fetch;
  
  t.after(() => {
    globalThis.fetch = originalFetch;
  });

  // Test Chat
  globalThis.fetch = async (url, options) => {
    assert.strictEqual(url, 'http://localhost:11434/api/chat');
    assert.strictEqual(options.method, 'POST');
    const body = JSON.parse(options.body);
    assert.strictEqual(body.model, 'gemma3');
    assert.strictEqual(body.messages[0].content, 'Hello');
    assert.strictEqual(body.stream, false);

    return {
      ok: true,
      json: async () => ({ message: { role: 'assistant', content: 'Hi!' } })
    };
  };
  const chatRes = await client.chat('gemma3', 'Hello');
  assert.strictEqual(chatRes.message.content, 'Hi!');

  // Test Generate
  globalThis.fetch = async (url, options) => {
    assert.strictEqual(url, 'http://localhost:11434/api/generate');
    const body = JSON.parse(options.body);
    assert.strictEqual(body.prompt, 'Write a poem');
    return {
      ok: true,
      json: async () => ({ response: 'Once upon a time...' })
    };
  };
  const genRes = await client.generate('gemma3', 'Write a poem');
  assert.strictEqual(genRes.response, 'Once upon a time...');

  // Test Compose
  globalThis.fetch = async (url, options) => {
    assert.strictEqual(url, 'http://localhost:11434/api/compose');
    const body = JSON.parse(options.body);
    assert.strictEqual(body.input, 'hello');
    assert.strictEqual(body.steps[0].model, 'gemma3');
    return {
      ok: true,
      json: async () => ({ output: 'composed output', results: [] })
    };
  };
  const compRes = await client.compose({
    input: 'hello',
    steps: [{ model: 'gemma3', prompt: 'test' }]
  });
  assert.strictEqual(compRes.output, 'composed output');

  // Test Messages (Anthropic compatible)
  globalThis.fetch = async (url, options) => {
    assert.strictEqual(url, 'http://localhost:11434/v1/messages');
    const body = JSON.parse(options.body);
    assert.strictEqual(body.max_tokens, 500);
    return {
      ok: true,
      json: async () => ({ content: [{ type: 'text', text: 'Anthropic reply' }] })
    };
  };
  const msgRes = await client.messages('gemma3', [{ role: 'user', content: 'hi' }], { maxTokens: 500 });
  assert.strictEqual(msgRes.content[0].text, 'Anthropic reply');

  // Test Chat Completions (OpenAI compatible)
  globalThis.fetch = async (url, options) => {
    assert.strictEqual(url, 'http://localhost:11434/v1/chat/completions');
    const body = JSON.parse(options.body);
    assert.strictEqual(body.temperature, 0.7);
    return {
      ok: true,
      json: async () => ({ choices: [{ message: { content: 'OpenAI reply' } }] })
    };
  };
  const compOpenAI = await client.chatCompletions('gemma3', [{ role: 'user', content: 'hi' }], { temperature: 0.7 });
  assert.strictEqual(compOpenAI.choices[0].message.content, 'OpenAI reply');

  // Test List Models
  globalThis.fetch = async (url, options) => {
    assert.strictEqual(url, 'http://localhost:11434/api/tags');
    return {
      ok: true,
      json: async () => ({ models: [{ name: 'gemma3' }, { name: 'llama3' }] })
    };
  };
  const models = await client.listModels();
  assert.strictEqual(models.length, 2);
  assert.strictEqual(models[0].name, 'gemma3');

  // Test Pull
  globalThis.fetch = async (url, options) => {
    assert.strictEqual(url, 'http://localhost:11434/api/pull');
    const body = JSON.parse(options.body);
    assert.strictEqual(body.model, 'gemma3');
    return {
      ok: true,
      json: async () => ({ status: 'success' })
    };
  };
  const pullRes = await client.pull('gemma3', { stream: false });
  assert.strictEqual(pullRes.status, 'success');

  // Test Show
  globalThis.fetch = async (url, options) => {
    assert.strictEqual(url, 'http://localhost:11434/api/show');
    const body = JSON.parse(options.body);
    assert.strictEqual(body.model, 'gemma3');
    return {
      ok: true,
      json: async () => ({ details: { parameter_size: '9B' } })
    };
  };
  const showRes = await client.show('gemma3');
  assert.strictEqual(showRes.details.parameter_size, '9B');

  // Test Is Running
  globalThis.fetch = async (url, options) => {
    assert.strictEqual(url, 'http://localhost:11434/');
    return {
      ok: true,
      json: async () => ({})
    };
  };
  const isRun = await client.isRunning();
  assert.strictEqual(isRun, true);
});

test('Lychee Client Error Handling', async (t) => {
  const client = new Lychee();
  const originalFetch = globalThis.fetch;
  
  t.after(() => {
    globalThis.fetch = originalFetch;
  });

  globalThis.fetch = async (url, options) => {
    return {
      ok: false,
      status: 404,
      text: async () => 'model not found'
    };
  };

  await assert.rejects(
    async () => {
      await client.chat('nonexistent', 'hi');
    },
    (err) => {
      assert.ok(err instanceof LycheeError);
      assert.strictEqual(err.statusCode, 404);
      assert.match(err.message, /HTTP 404: model not found/);
      return true;
    }
  );
});

File: sdk/python/tests/test_client.py
import unittest
from unittest.mock import patch, MagicMock
import json

from lychee import Client

class TestLycheeClient(unittest.TestCase):
    def setUp(self):
        self.client = Client()

    @patch('requests.request')
    def test_generate(self, mock_request):
        mock_response = MagicMock()
        mock_response.json.return_value = {"model": "llama3", "response": "hello", "done": True}
        mock_response.status_code = 200
        mock_request.return_value = mock_response

        res = self.client.generate("llama3", "hi")
        self.assertEqual(res["response"], "hello")
        mock_request.assert_called_once_with(
            "POST",
            "http://localhost:11434/api/generate",
            json={"model": "llama3", "prompt": "hi", "stream": False, "raw": False}
        )

    @patch('requests.request')
    def test_generate_stream(self, mock_request):
        mock_response = MagicMock()
        mock_response.iter_lines.return_value = [
            b'{"response": "hel", "done": false}',
            b'{"response": "lo", "done": true}'
        ]
        mock_response.status_code = 200
        mock_request.return_value = mock_response

        generator = self.client.generate("llama3", "hi", stream=True)
        chunks = list(generator)
        self.assertEqual(len(chunks), 2)
        self.assertEqual(chunks[0]["response"], "hel")
        self.assertEqual(chunks[1]["response"], "lo")

    @patch('requests.request')
    def test_chat(self, mock_request):
        mock_response = MagicMock()
        mock_response.json.return_value = {"model": "llama3", "message": {"role": "assistant", "content": "hi"}}
        mock_response.status_code = 200
        mock_request.return_value = mock_response

        messages = [{"role": "user", "content": "hello"}]
        res = self.client.chat("llama3", messages)
        self.assertEqual(res["message"]["content"], "hi")

    @patch('requests.request')
    def test_chat_stream(self, mock_request):
        mock_response = MagicMock()
        mock_response.iter_lines.return_value = [
            b'{"message": {"role": "assistant", "content": "h"}, "done": false}',
            b'{"message": {"role": "assistant", "content": "i"}, "done": true}'
        ]
        mock_response.status_code = 200
        mock_request.return_value = mock_response

        messages = [{"role": "user", "content": "hello"}]
        generator = self.client.chat("llama3", messages, stream=True)
        chunks = list(generator)
        self.assertEqual(len(chunks), 2)
        self.assertEqual(chunks[0]["message"]["content"], "h")
        self.assertEqual(chunks[1]["message"]["content"], "i")

    @patch('requests.request')
    def test_list(self, mock_request):
        mock_response = MagicMock()
        mock_response.json.return_value = {"models": [{"name": "llama3:latest"}]}
        mock_response.status_code = 200
        mock_request.return_value = mock_response

        res = self.client.list()
        self.assertEqual(len(res["models"]), 1)
        self.assertEqual(res["models"][0]["name"], "llama3:latest")

    @patch('requests.request')
    def test_show(self, mock_request):
        mock_response = MagicMock()
        mock_response.json.return_value = {"modelfile": "FROM llama3"}
        mock_response.status_code = 200
        mock_request.return_value = mock_response

        res = self.client.show("llama3")
        self.assertEqual(res["modelfile"], "FROM llama3")

    @patch('requests.request')
    def test_create(self, mock_request):
        mock_response = MagicMock()
        mock_response.json.return_value = {"status": "success"}
        mock_response.status_code = 200
        mock_request.return_value = mock_response

        res = self.client.create("my-model", modelfile="FROM llama3")
        self.assertEqual(res["status"], "success")

    @patch('requests.request')
    def test_delete(self, mock_request):
        mock_response = MagicMock()
        mock_response.json.return_value = {"status": "deleted"}
        mock_response.status_code = 200
        mock_request.return_value = mock_response

        res = self.client.delete("my-model")
        self.assertEqual(res["status"], "deleted")

    @patch('requests.request')
    def test_pull(self, mock_request):
        mock_response = MagicMock()
        mock_response.json.return_value = {"status": "pulling"}
        mock_response.status_code = 200
        mock_request.return_value = mock_response

        res = self.client.pull("llama3")
        self.assertEqual(res["status"], "pulling")

    @patch('requests.request')
    def test_push(self, mock_request):
        mock_response = MagicMock()
        mock_response.json.return_value = {"status": "pushing"}
        mock_response.status_code = 200
        mock_request.return_value = mock_response

        res = self.client.push("llama3")
        self.assertEqual(res["status"], "pushing")

    @patch('requests.request')
    def test_embed(self, mock_request):
        mock_response = MagicMock()
        mock_response.json.return_value = {"embeddings": [[0.1, 0.2]]}
        mock_response.status_code = 200
        mock_request.return_value = mock_response

        res = self.client.embed("llama3", "hello")
        self.assertEqual(len(res["embeddings"]), 1)
        self.assertEqual(res["embeddings"][0], [0.1, 0.2])

    @patch('requests.request')
    def test_ps(self, mock_request):
        mock_response = MagicMock()
        mock_response.json.return_value = {"models": []}
        mock_response.status_code = 200
        mock_request.return_value = mock_response

        res = self.client.ps()
        self.assertEqual(res["models"], [])

if __name__ == '__main__':
    unittest.main()
